### PR TITLE
[quant] Fix rigor gate root causes: degenerate series, look-ahead false positives, stale leaderboard (#868)

### LINKED ISSUE #868: [quant] Rigor gate 0/31 root cause: degenerate series poison the denominator + look-ahead false-positives + leaderboard reads stale fixtures
## Summary
Dogfooding (#854 finding 3) showed **0 of 31 library strategies pass the rigor gate** on the live site, killing the generate→gate→deploy spine for every visitor. A root-cause investigation (2026-07-03, live API + code replay) found this is **three distinct problems compounding**, not one bug — and the gate is partly *right*:

1. **14 strategies have real, non-degenerate returns and honestly fail DSR p≥0.95** (p range 0.003–0.85; holds even at num_trials=1). Not a bug — possibly a framing/copy question (Önder's call, see #822/#819 context).
2. **11 strategies carry zero-variance daily_returns** (placeholder-flat series): DSR/OOS correctly return None — but these degenerate series still **inflate `num_trials` (25 vs ~14 true)** and **dilute `avg_correlation`**, unfairly stiffening DSR for the 14 real ones (`selection_bias_routes.py:176-183` filters only `len(v)>=10`, no variance check).
3. **6 strategies show nonzero Sharpe/CAGR on the leaderboard with ZERO persisted daily_returns** (`is_backtest_placeholder=false` despite no rows) — a data-path bug (id mismatch suspected, esp. the 4 duplicate Pairs Trading entries; leaderboard reads stale `s.dsr_p_value`/fixture fields at `strategies_routes.py:117-135` while only `passes_rigor_gate` uses the live verdict).

Plus a **look-ahead false-positive class**: `look_ahead_audit` flags ANY negative-index subscript, but backtrader's `self.data.close[-i]` is bars-AGO access (e.g. `moreira_muir_2017_volatility_managed.py:99-100`) — standard, not look-ahead — and `passes_all` requires zero warnings → hard FAIL.

## Scope
- `backend/archimedes/api/selection_bias_routes.py` — exclude zero-variance series from `valid_returns` BEFORE computing num_trials / avg_correlation / cohort PBO (mirror `_rigor_helpers._sharpe_dsr_inputs`'s `np.ptp(arr)==0` test).
- `backend/archimedes/services/rigor_evaluator.py` (`look_ahead_audit`) — skip/whitelist backtrader `self.data.*[-i]` bars-ago access; keep genuine future-index detection.
- `backend/archimedes/api/strategies_routes.py:117-135` — serve the LIVE gate verdict's numbers (dsr_p_value/pbo/oos), not stale fixture fields, so the leaderboard can't disagree with the gate.
- Investigate + backfill/mark-ungateable the 6 zero-daily-returns ids and the 11 zero-variance series (data task; coordinate with #863/#864's fixtures migration).

## Acceptance criteria
- [ ] `GET /api/selection-bias/gate`: `num_trials` counts only non-degenerate series (expect ~14 with today's data).
- [ ] `moreira_muir_2017_volatility_managed` no longer fails look-ahead on `close[-i]`; a synthetic true-look-ahead fixture (positive future index) still FAILS.
- [ ] Leaderboard metrics for any id equal the live gate's values for that id (spot-check 5 ids incl. one of the 6 broken ones).
- [ ] Hermetic tests for each change; no thresholds weakened (DSR p≥0.95, PBO<0.5 untouched).

## Anti-goals
- Do NOT relax any gate threshold to make strategies pass — if the honest 14 still fail after the denominator/audit fixes, that's a product-framing decision (separate discussion).
- Do NOT touch `_rigor_helpers` math without Önder's sign-off (his lane; precedent #811/#817).

Full evidence trail (file:line for every claim) in the 2026-07-03 investigation — ask Dan/Claude for the transcript.

### PR DESCRIPTION
**Pull request by onder-akkaya**
## Summary

Closes #868. The live rigor gate was passing 0 of 31 library strategies. Not because the library is all bad, three separate bugs were feeding the gate bad inputs, plus one of its own checks had a false-positive problem. The 14 strategies that genuinely fail DSR honestly still fail after this PR, that part of the gate was working correctly the whole time.

## What changed

- `backend/archimedes/api/selection_bias_routes.py`: `valid_returns` now excludes zero-variance (flat/placeholder) daily-return series before computing `num_trials`, `avg_correlation`, and cohort PBO. 11 of the 31 library entries are placeholder stubs with no real backtest behind them; counting them toward the multiple-testing correction was unfairly stiffening DSR for the 14 real strategies with a genuine shot at passing. Mirrors the same degeneracy test (`np.ptp(arr) == 0`) `_rigor_helpers._sharpe_dsr_inputs` already uses, so "degenerate" means the same thing everywhere in the gate. Every strategy, degenerate or not, still runs its own gate check individually and reports its own honest verdict; this only trims the cohort-level context.
- `backend/archimedes/services/rigor_evaluator.py`, `look_ahead_audit`: negative-index subscripts on backtrader line-buffer access (`self.data.close[-i]`, a `self.datas` loop-variable alias, a self-rooted custom indicator) no longer trigger a false positive. `self.data.close[-i]` is backtrader's standard "i bars ago" convention, not a look-ahead violation, but the old check flagged it anyway and the gate requires zero warnings to pass. A pandas `.iloc`/`.loc`/`.at`/`.iat` accessor, or a negative index on anything that isn't backtrader line access, still gets flagged. The positive-index branch (genuine future-index detection) is untouched.
- `backend/archimedes/api/strategies_routes.py`: the leaderboard's numeric fields (`dsr_p_value`, `pbo_score`, `out_of_sample_sharpe`, `deflated_sharpe_ratio`) now come from the same live gate run that already produces the `passes_rigor_gate` badge (#821), instead of stale fixture-derived fields. Previously only the boolean badge read the live verdict while the numbers next to it could show something else entirely for the same strategy. Falls back to the old fixture chain when the live gate can't run (no persisted returns, DB failure), fail-closed, no fabricated numbers.

## Verify

```
env -i HOME=$HOME PATH=$PATH PYTHONPATH=backend python -m pytest backend/tests/test_rigor_evaluator.py backend/tests/test_selection_bias_routes.py backend/tests/test_strategies_routes.py -q   # 259 passed
env -i HOME=$HOME PATH=$PATH PYTHONPATH=backend python -m pytest -m "not integration" -q   # 2374 passed
```

## Found but deliberately not fixed here

The look-ahead audit's positive-index branch has its own, separate false positive: `self.datas[N]` (a constant selecting which data feed, standard in a multi-asset strategy) gets flagged as "may reference future bars," conflating feed selection with a time offset. This is present on unmodified `main`, independent of anything in this PR, and is currently why `gatev_2006_pairs_distance.py`, `elliott_2005_kalman_pairs.py`, and `engle_granger_1987_cointegration_pairs.py` still fail look-ahead after this lands, for a different reason than #868 asked me to fix. Left the positive-index detector untouched per #868's own anti-goal against touching it casually. Filed #881 to track it separately.

## Anti-goals checked

- No gate threshold changed. DSR p≥0.95 and PBO<0.5 are exactly what they were.
- `_rigor_helpers.py`'s DSR/PBO math is untouched. This PR only changes what feeds it and what the leaderboard displays.
- Did not touch the positive-index look-ahead detector (see above, #881).

## One acceptance criterion I couldn't verify hermetically

#868 asks to spot-check 5 real strategy ids, including one of the "6 broken ones," against live data. This sandbox has no persisted `strategy_daily_returns` rows to check against, so I wrote the parity test as "the leaderboard route serves the same values `selection_bias_routes` computes for a given id" against fixture data instead (`test_leaderboard_never_disagrees_with_selection_bias_gate_route`), which passes. Live-id spot-checking is a follow-up for whoever has prod DB access, same shape as the #862/#864 backfill gap.

## Cross-lane note

Touches the rigor gate's routing/evaluator layer, which #822/#819/#817 also live in. No overlap with those PRs' files. Requesting @dbrowneup's review given the live-gate surface this touches and his #868 investigation.

### CODE REVIEW

File: backend/archimedes/api/strategies_routes.py [hunk 1]
```diff
@@ -217,8 +417,9 @@ async def list_strategies(
     total = len(strats)
     window = strats[offset : offset + limit]
     verdicts = verdicts_for_strategies(window)
+    rigor_results = _live_rigor_results_for_strategies(window)
     return StrategyListResponse(
-        strategies=[_to_strategy_response(s, verdicts.get(s.id)) for s in window],
+        strategies=[_to_strategy_response(s, verdicts.get(s.id), rigor_results.get(s.id)) for s in window],
         total=total,
```
[the remark this answers is not in the document]
Comment (dbrowneup) [reply backend/archimedes/api/strategies_routes.py]: Fixed in 71dabb7 — added `_verdict_from_result(result)` helper that derives `RigorGateVerdict` from an already-computed `RigorGateResult`, and updated `list_strategies` to use it instead of the separate `verdicts_for_strategies(window)` call. One DB read + one gate run per request now. The cohort-filtering inconsistency (degenerate-series exclusion was present in `_live_rigor_results_for_strategies` but not in `verdicts_for_strategies`) is eliminated because badge and numeric fields now come from the same computation.

File: backend/archimedes/api/strategies_routes.py [hunk 2]
```diff
@@ -189,6 +230,158 @@ def _live_verdict_for_one(s: Strategy) -> RigorGateVerdict:
...
+            results[s.id] = run_rigor_gate(
+                strategy_id=s.id,
+                daily_returns=daily_returns,
+                num_trials=num_trials,
+                pbo_scores=pbo_scores,
+                strategy_code=code_by_id.get(s.id),
+                in_sample_sharpe=None,
+                paper_claimed_sharpe=getattr(s, "paper_claimed_sharpe", None),
+                average_correlation=avg_correlation,
+            )
```
[the remark this answers is not in the document]
Comment (dbrowneup) [reply backend/archimedes/api/strategies_routes.py]: Fixed in 71dabb7 — the `if s.id not in valid_returns: continue` guard was removed. `_live_rigor_results_for_strategies` now iterates all strategies with >= 10 returns: non-degenerate series (in `valid_returns`) get the full cohort context (cohort `num_trials`, `pbo_scores`, `avg_correlation`); degenerate (zero-variance) series get `num_trials=1, pbo_scores={}, average_correlation=0.0` — excluded from cohort to avoid inflating num_trials/diluting avg_correlation, but their own gate still runs and produces a live 'fail' result. Strategies with < 10 returns are still absent from the dict (genuine no-data case). `code_by_id` was updated to load source for all strategies with >= 10 returns, not just `valid_returns`.

File: backend/archimedes/services/rigor_evaluator.py [hunk 3]
```diff
@@ -61,6 +61,69 @@
...
+
+    if any(a in _PANDAS_ACCESSOR_ATTRS for a in attrs):
+        return False
+    if root == "self":
+        return True
+    # attrs[0] is the innermost-collected == outermost/subscripted attribute
+    # (e.g. for `d.close`, attrs == ["close"]; for `self.data.close`,
+    # attrs == ["close", "data"]) — either way the subscripted attribute is
+    # attrs[0].
+    return bool(attrs) and attrs[0] in _BACKTRADER_LINE_NAMES
```
[the remark this answers is not in the document]
Comment (dbrowneup) [reply backend/archimedes/services/rigor_evaluator.py:125]: Won't fix — the non-`self` whitelist in `_is_backtrader_line_access` is specifically designed for `for d in self.datas: ... d.close[-i]` loop-variable patterns, which are backtrader's standard idiom and the only realistic scenario where a root that isn't literally `self` carries a backtrader OHLCV attribute name subscripted with a negative index. The `df.close[-1]` case Copilot raises is technically possible but extremely rare in real pandas code — conventional pandas patterns for last-row access are `df.iloc[-1]`, `df['close'].iloc[-1]`, or `df.close.iloc[-1]`, all of which ARE caught (iloc/loc/at/iat in `_PANDAS_ACCESSOR_ATTRS`, or the negative-index-on-non-attribute guard). Tightening the non-self branch to require a single-char root would break `data.close[-i]` (common alias for `self.datas[0]`); tightening it to require the root appear in a known set of loop-var names would be fragile and wrong on arbitrary strategy code. The static audit's role is reviewer-support, not airtight enforcement — a `df.close[-1]` false-negative is a low-probability edge case in a backtrader strategy, and the heuristic's false-positive cost (flagging legitimate bars-ago access and blocking valid strategies) is higher. Verified: all existing look-ahead tests pass (259 passed, 0 failed).
